src/irmin-test/layered_store.ml

```diff
@@ -0,0 +1,958 @@
+              S.Branch.set repo b c >>= fun () ->
+              setup (i + 1) tree (commits @ [ c ]) contents
+          | 3 ->
+              S.Commit.v repo ~info
+                ~parents:[ S.Commit.hash (List.nth commits 1) ]
+                tree'
+              >>= fun c ->
+              S.Branch.set repo b c >>= fun () ->
+              setup (i + 1) tree' (commits @ [ c ]) contents
+          | 4 ->
+              S.Commit.v repo ~info
+                ~parents:[ S.Commit.hash (List.nth commits 3) ]
+                tree'
+              >>= fun c -> setup (i + 1) tree (commits @ [ c ]) contents
+          | 5 | 6 ->
+              S.Commit.v repo ~info
+                ~parents:[ S.Commit.hash (List.nth commits 3) ]
+                tree'
+              >>= fun c ->
+              S.Branch.set repo b c >>= fun () ->
+              setup (i + 1) tree (commits @ [ c ]) contents
+          | _ -> assert false
+      in
+      setup 0 tree [] [] >>= fun (commits, contents) ->
+      S.freeze repo ~min:[ List.nth commits 1 ] ~max:[ List.nth commits 3 ]
+        ~copy_in_upper:true
+      >>= fun () ->
+      S.PrivateLayer.wait_for_freeze () >>= fun () ->
+      S.freeze repo ~min:[ List.nth commits 1 ] ~max:[ List.nth commits 3 ]
+        ~copy_in_upper:true
+      >>= fun () ->
+      S.PrivateLayer.wait_for_freeze () >>= fun () ->
+      Lwt_list.iter_p
+        (fun i ->
+          P.Commit.find (h repo) (S.Commit.hash (List.nth commits i))
+          >|= fun t -> check_commit "deleted commit" None t)
+        [ 0; 2; 4 ]
+      >>= fun () ->
+      S.Branch.find repo "b2" >>= fun c2' ->
+      check_branch repo "b2" None c2';
+      S.Branch.find repo "b3" >>= fun c3' ->
+      check_branch repo "b3" (Some (List.nth commits 3)) c3';
+      S.Branch.find repo "b5" >>= fun c5' ->
+      check_branch repo "b5" (Some (List.nth commits 5)) c5';
+      S.Branch.find repo "b6" >>= fun c6' ->
+      check_branch repo "b6" (Some (List.nth commits 6)) c6';
+      check_layer_id repo
+        (S.Commit_t (S.Commit.hash (List.nth commits 1)))
+        "layer id of c1" `Lower
+      >>= fun () ->
+      Lwt_list.iter_p
+        (fun i ->
+          check_layer_id repo
+            (S.Commit_t (S.Commit.hash (List.nth commits i)))
+            ("layer id of c" ^ string_of_int i)
+            `Upper)
+        [ 3; 5; 6 ]
+      >>= fun () ->
+      S.of_branch repo "b6" >>= fun t ->
+      S.get_tree t [] >>= fun tree ->
+      fail_with_none (S.Tree.find tree [ "a"; "b"; "c0" ]) "find x0"
+      >>= fun s ->
+      Alcotest.(check string) "x0" s "x0";
+      fail_with_none (S.Tree.find tree [ "a"; "b"; "c1" ]) "find x1"
+      >>= fun s ->
+      Alcotest.(check string) "x1" s "x1";
+      Lwt_list.iter_p
+        (fun i ->
+          check_layer_id repo
+            (S.Content_t (List.nth contents i))
+            ("layer id of x" ^ string_of_int i)
+            `Upper)
+        [ 0; 1; 3 ]
+      >>= fun () ->
+      Lwt.catch
```

Comment:
[This helper function](https://github.com/mirage/irmin/blob/39454c5240adef463c1a37b7e0cf267827c1078a/test/irmin-pack/common.ml#L109) might be useful here and in a few other places.